Engine/GameEngine.cs

```diff
@@ -0,0 +1,410 @@
+        int topUpPileSize = combinedCenterPiles.Count / gameState.Players.Count;
+        foreach (Player player in gameState.Players)
+        {
+            player.TopUpCards = combinedCenterPiles.PopRange(topUpPileSize);
+        }
+
+        // Reset the center piles
+        var newCenterPiles = new List<List<Card>>();
+        for (var i = 0; i < gameState.CenterPiles.Count; i++) newCenterPiles.Add(new List<Card>());
+        gameState.CenterPiles = newCenterPiles;
+
+        return new SuccessResult<GameState>(gameState);
+    }
+
+    public static Result<GameState> TryPlayCard(GameState gameState, Player player, Card card,
+        int centerPileIndex)
+    {
+        if (centerPileIndex >= gameState.CenterPiles.Count)
+        {
+            return new ErrorResult<GameState>($"No center pile found at index {centerPileIndex}");
+        }
+
+        Result<CardLocation> cardLocationResult = FindCardLocation(gameState, card);
+        if (cardLocationResult is IErrorResult cardLocationResultError)
+        {
+            return new ErrorResult<GameState>(cardLocationResultError.Message);
+        }
+
+        if (cardLocationResult.Data.PlayerIndex != gameState.Players.IndexOf(player))
+        {
+            return new ErrorResult<GameState>($"Player {player.Name} does not have card {card.Value} in their hand");
+        }
+
+        switch (cardLocationResult.Data.PileName)
+        {
+            case CardPileName.Hand:
+                // Check that the card can be played onto the relevant center piles top card
+                if (!ValidMove(card, gameState.CenterPiles[centerPileIndex].Last()))
+                {
+                    return new ErrorResult<GameState>(
+                        $"Card with value {card.Value} can't be played onto {gameState.CenterPiles[centerPileIndex].Last().Value}");
+                }
+
+                // This is a valid play so play it
+                return PlayCard(gameState, card, centerPileIndex);
+
+            default:
+                return new ErrorResult<GameState>(
+                    $"Can't play a card from the {cardLocationResult.Data.PileName} pile");
+        }
+    }
+
+    private static Result<GameState> PlayCard(GameState gameState, Card card, int centerPileIndex)
+    {
+        gameState.CenterPiles[centerPileIndex].Add(card);
+        Player? playerWithCard = gameState.Players.FirstOrDefault(p => p.HandCards.Contains(card));
+        playerWithCard?.HandCards.Remove(card);
+
+        // Reset any ones request to top up if they can now move
+        foreach (Player player in gameState.Players)
+        {
+            if (player.RequestingTopUp && PlayerHasPlay(gameState, player).Success)
+            {
+                player.RequestingTopUp = false;
+            }
+        }
+
+        return new SuccessResult<GameState>(gameState);
+    }
+
+    public static Result<(GameState updatedGameState, Card pickedUpCard)> TryPickupFromKitty(
```

Comment:
I see alot of usage of tuples for return types, I think it can potentially read easier if these were there own seperate classes e.g. here `KittyAction` or `KittyResult` or something along those lines.